### DISCUSSION TITLE
v4.0.0-beta.1 missing Qt project file - cannot compile transmission-qt client

### DISCUSSION THREAD

**Discussion by admusin** [transmission/transmission#4102] [Support]
In the `qt/` folder, the project file (`qt.pro`) is missing for **v4.0.0-beta.1**.
Since for Linux, all we get are the sources inside the tarball or in the Git repo, it's near to impossible to compile the `transmission-qt` executable for those of us who are not experienced devs (i.e. without knowledge about how to generate the proper project file in general and in particular for this project).

Is it missing on purpose (e.g. the Qt code for this version is not yet ready to release)?
If not, would it possible to add it, so we can build the Qt GUI as well, for this beta?

**TL;DR:** I'd like to be able to compile `transmission-qt` from the new beta sources.

P.S. I also posted this as a comment on the beta commit 42713c6716e748031d5de4a4e0f94b28b7ec0484 but since I'm not sure if it's visible there, I opened this discussion.

**Comment by fredo-47** [comment]
I am also on Linux, and I have no problems compiling `transmission-qt` from the nightly sources.
Just follow the steps listed e.g. at "[Building Transmission from Git](https://github.com/transmission/transmission#building-transmission-from-git-first-time)".

PS: I don't know what a "project file`qt.pro`" would be required for.

**Comment by ile6695** [comment]
You're likely missing some dependencies for building the qt client. I'm assuming you want qt 5 client and not qt 6.
Check that you have `qttools5-dev` installed.

The instruction for building from git is the same as building from tarball.
Only difference is that you download the tarball instead of cloning git repository.

**Comment by admusin** [comment]
@fredo-47 
I see, I took a look in `CMakeLists.txt` and by default it seems that it's supposed to also build the GUI clients, provided it finds the needed Qt/GTK packages installed.
The thing is that for me it didn't; the only binaries I got inside the build folder at the end, were:
```
~/Downloads/transmission-4.0.0-beta.1+r98cf7d9b3c/build$ find . -name transmission* -type f
./daemon/transmission-daemon
./utils/transmission-edit
./utils/transmission-remote
./utils/transmission-show
./utils/transmission-create
```
Seeing that I've gotten no binaries for the GUI clients after the build was done, I assumed they need to be built separately, and I went in the `qt` folder and tried to follow the build instructions from the README.txt file; those ones DO need you to have a `.pro` file in there, as it's expected by `qmake` to help it generate the makefile or something. Therefore, when you run `qmake`, as per the README.txt file, you get an error because it's missing.

Nevermind, back to the default build instructions.

@ile6695
Indeed, I'm running Plasma 5.20.5 with Qt 5.15.2 so I'm targeting **Qt 5**. 
And you were spot on; `qttools5-dev` was not installed on the machine where I build (a Debian 11) and after installing it and scratching the whole `build` folder, then running again `cmake` and `make`, I finally got the GUI binaries built.

The weird part is that neither before nor after installing `qttools5-dev` did I ever get an error or warning in the output of `cmake` or `make`, saying anything at all about `Qt`packages missing.

I guess the issue lies somehow in the CMakeLists.txt file, as I can see there is in there a line `find_package(Qt NAMES ${Qt_NAMES} ${QT_MINIMUM} QUIET)` and I assume that `QUIET` may mean that it won't complain if it doesn't find something it needs.

Lower after the above statement there's:
```
    if(NOT QT_TARGETS OR NOT (ENABLE_QT_COM_INTEROP OR ENABLE_QT_DBUS_INTEROP))
        if(QT_IS_REQUIRED)
            message(FATAL_ERROR "Unable to find required Qt libraries (Qt${Qt_VERSION_MAJOR}${MISSING_QT_MODULE})")
        endif()
        set(QT_FOUND OFF)
    endif()

```
which seems to suggest that one should see an error line `Unable to find required Qt libraries...` appearing in the `cmake` output if something was wrong with the required Qt packages but I never got this error before installing `qttools5-dev`.
I even `tee`-ed the `cmake` output to a file to search it easier and make sure I don't miss out any errors.
So, I guess the Qt prerequisites detection mechanism in the file is not entirely foolproof. 
Alas, it's above my skill-set to debug it, so I can't contribute to improving it right now.

That being said, thank you both for replying and steering me in the right direction.

**Reply by mikedld** [reply]
You're missing an important part in your analysis of the quoted piece of CMake code: the `if(QT_IS_REQUIRED)` check. Default value for the `ENABLE_QT` option is `AUTO` which results in Qt client being built when dependency requirements are satisfied, and ignored otherwise. If you were to pass `-DENABLE_QT=ON` to CMake when configuring the project, thus explicitly marking the Qt client as required, you'd see the message.

**Comment by ile6695** [comment]
It's not giving errors because cmake thinks that because you don't have required libraries, you likely don't want to build that version. In the beginning of `CMakeLists.txt` you can find many optional things, such as:

```
tr_auto_option(ENABLE_GTK           "Build GTK client" AUTO)
tr_auto_option(ENABLE_QT            "Build Qt client" AUTO)
tr_auto_option(ENABLE_MAC           "Build Mac client" AUTO)
```
You can force these to be enabled by adding cmake variables, for example
```
cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_QT=ON ..
```
and then it will fail if something is missing.

Funny coincidence that I made a pr that would update Debian 11 build instructions 3 hours before you made this thread. I wasn't expecting someone to need it so soon.

**Comment by admusin** [comment]
Hi @mikedld , @ile6695,
I take back what I had said above about the detection mechanism in CMakeLists.txt not doing it's job.
After reading your posts above, I realize that it was my lack of knowledge regarding the file's logic that made me think so.

I've really, restarted my C++ learning journey only a few days ago, which will likely take me through `make` and `cmake` land eventually but at present, this still feels like arcane knowledge to me.

Following the reply from @ile6695, earlier, I managed to solve it but even more importantly, now I've learned something new.
Thank you all for getting back at me and helping me sort this out.

**Comment by ckerr** [comment]
At this point we should probably remove `qt/README.txt`
